### PR TITLE
[SPARK-22462][SQL] Make rdd-based actions in Dataset trackable in SQL UI

### PR DESCRIPTION
**Pull request by viirya**
## What changes were proposed in this pull request?

For the few Dataset actions such as `foreach`, currently no SQL metrics are visible in the SQL tab of SparkUI. It is because it binds wrongly to Dataset's `QueryExecution`. As the actions directly evaluate on the RDD which has individual `QueryExecution`, to show correct SQL metrics on UI, we should bind to RDD's `QueryExecution`.

## How was this patch tested?

Manually test. Screenshot is attached in the PR.

### CODE REVIEW

**Comment by viirya** [comment]
The screenshot for running `sql("select * from range(10)").foreach(a => Unit)` on spark-shell:

screen shot 2017-11-08 at 12 07 37 pm

**Comment by viirya** [comment]
retest this please.

**Comment by viirya** [comment]
cc @juliuszsompolski

**Comment by juliuszsompolski** [comment]
Thanks for the fix @viirya! But I'm not a Spark committer to approve it :-).

**Comment by viirya** [comment]
@juliuszsompolski No problem. Non-committer can still review. :)

**Comment by viirya** [comment]
cc @cloud-fan for review too.

**Comment by HyukjinKwon** [comment]
and also I believe anyone can leave the sign-off too if it looks good :).

File: sql/core/src/main/scala/org/apache/spark/sql/Dataset.scala [hunk 1]
```diff
@@ -2579,8 +2579,15 @@ class Dataset[T] private[sql](
    * @group action
    * @since 1.6.0
    */
-  def foreach(f: T => Unit): Unit = withNewExecutionId {
```
Comment (cloud-fan) [inline sql/core/src/main/scala/org/apache/spark/sql/Dataset.scala]: how about we add a `withNewRDDExecutionId`?

Comment (viirya) [reply sql/core/src/main/scala/org/apache/spark/sql/Dataset.scala]: Ok.

File: sql/core/src/main/scala/org/apache/spark/sql/Dataset.scala [hunk 2]
```diff
@@ -2579,8 +2579,10 @@ class Dataset[T] private[sql](
...
-    rdd.foreach(f)
+  def foreach(f: T => Unit): Unit = {
+    withNewRDDExecutionId {
```
Comment (cloud-fan) [inline sql/core/src/main/scala/org/apache/spark/sql/Dataset.scala]: we can follow the previous style: `def foreach ... = withNewRDDExecutionId  {`

**Comment by cloud-fan** [comment]
LGTM

**Comment by cloud-fan** [comment]
thanks, merging to master!

**Pull request closed without merging**